shell/platform/windows/accessibility_root_node.cc

```diff
@@ -0,0 +1,287 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/accessibility_root_node.h"
+
+#include "flutter/fml/logging.h"
+#include "flutter/third_party/accessibility/base/win/atl_module.h"
+
+namespace flutter {
+
+static constexpr LONG kWindowChildId = 1;
+static constexpr LONG kInvalidChildId = 3;
+
+AccessibilityRootNode::AccessibilityRootNode() : alert_accessible_(nullptr) {}
+
+AccessibilityRootNode::~AccessibilityRootNode() {
+  if (alert_accessible_) {
+    alert_accessible_->Release();
+    alert_accessible_ = nullptr;
+  }
+}
+
+IAccessible* AccessibilityRootNode::GetTargetAndChildID(VARIANT* var_id) {
+  LONG& child_id = var_id->lVal;
+  if (V_VT(var_id) != VT_I4) {
+    child_id = kInvalidChildId;
+    return nullptr;
+  }
+  child_id = V_I4(var_id);
+  if (!window_accessible_) {
+    return nullptr;
+  }
+  if (child_id == CHILDID_SELF || child_id == kWindowChildId) {
+    child_id = CHILDID_SELF;
+    return window_accessible_;
+  }
+  if (child_id == kAlertChildId && alert_accessible_) {
+    child_id = CHILDID_SELF;
+    return alert_accessible_;
+  }
+  // A negative child ID can be used to refer to an AX node directly by its ID.
+  if (child_id < 0) {
+    return window_accessible_;
+  }
+  return nullptr;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accHitTest(LONG screen_physical_pixel_x,
+                                                 LONG screen_physical_pixel_y,
+                                                 VARIANT* child) {
+  if (window_accessible_) {
+    return window_accessible_->accHitTest(screen_physical_pixel_x,
+                                          screen_physical_pixel_y, child);
+  }
+  child->vt = VT_EMPTY;
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accDoDefaultAction(VARIANT var_id) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
```

Comment:
nit: I'm assuming our lint rules force the extra brackets here? If not, delete them.

Reply:
That is correct, they are treated as a "did you mean `==`?" error otherwise.